Objects/genericaliasobject.c

```diff
@@ -416,21 +427,43 @@ _Py_subs_parameters(PyObject *self, PyObject *args, PyObject *parameters, PyObje
         t = list[T];          t[int]      -> newargs = [int]
         t = dict[str, T];     t[int]      -> newargs = [str, int]
         t = dict[T, list[S]]; t[str, int] -> newargs = [str, list[int]]
+        t = list[[T]];        t[str]      -> newargs = [[str]]
      */
-    Py_ssize_t nargs = PyTuple_GET_SIZE(args);
+    assert (PyTuple_Check(args) || PyList_Check(args));
+    Py_ssize_t nargs = PySequence_Length(args);
     PyObject *newargs = PyTuple_New(nargs);
     if (newargs == NULL) {
         Py_DECREF(item);
         return NULL;
     }
     for (Py_ssize_t iarg = 0, jarg = 0; iarg < nargs; iarg++) {
-        PyObject *arg = PyTuple_GET_ITEM(args, iarg);
+        PyObject *arg = PyTuple_Check(args) ?
+            PyTuple_GET_ITEM(args, iarg) : PyList_GET_ITEM(args, iarg);
         if (PyType_Check(arg)) {
             PyTuple_SET_ITEM(newargs, jarg, Py_NewRef(arg));
             jarg++;
             continue;
         }
-
+        // Recursively substitute params in lists/tuples.
+        if (PyTuple_Check(arg) || PyList_Check(arg)) {
+            PyObject *subargs = _Py_subs_parameters(self, arg, parameters, item);
+            if (subargs == NULL) {
+                Py_DECREF(newargs);
+                Py_DECREF(item);
+                return NULL;
+            }
+            PyObject *subargs_list = PySequence_List(subargs);
+            Py_DECREF(subargs);
+            if (subargs_list == NULL) {
+                Py_DECREF(newargs);
+                Py_DECREF(item);
+                return NULL;
+            }
+            Py_SETREF(subargs, subargs_list);
+            PyTuple_SET_ITEM(newargs, jarg, subargs);
```

Comment:
```suggestion
            PyTuple_SET_ITEM(newargs, jarg, subargs_list);
```
This fixes the crash for me locally. I believe your code double-decrefed `subargs`, and then later another tuple got allocated at the same place and fun things happened.

Reply:
Nice catch, thanks! You saved me a lot of time trying to debug this :sweat_smile:

Reply:
For what it's worth, my technique was to use `lldb` to reproduce the crash (`lldb python.exe`, `r -m test test_genericalias`; depending on your OS you will want a different debugger); then use `bt` to get the stack trace; then stare at the code to find possible culprits.

I also had the vague idea that the crash error probably indicates some refcounting issue, since I think it means that we try to GC-track a new tuple at a place where a previous tuple still thinks it lives.